azure/services/privatedns/spec.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package privatedns
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/privatedns/mgmt/2018-09-01/privatedns"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/converters"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+// ZoneSpec defines the specification for private dns zone.
+type ZoneSpec struct {
+	Name           string
+	ResourceGroup  string
+	ClusterName    string
+	AdditionalTags infrav1.Tags
+}
+
+// LinkSpec defines the specification for a virtual network link in a private DNS zone.
+type LinkSpec struct {
+	Name              string
+	ZoneName          string
+	SubscriptionID    string
+	VNetResourceGroup string
+	VNetName          string
+	ResourceGroup     string
+	ClusterName       string
+	AdditionalTags    infrav1.Tags
+}
+
+type RecordSpec struct {
+	Record        infrav1.AddressRecord
+	ZoneName      string
+	ResourceGroup string
+}
+
+// ResourceName returns the name of the private dns zone.
+func (z ZoneSpec) ResourceName() string {
+	return z.Name
+}
+
+// OwnerResourceName is a no-op for private dns zone.
+func (z ZoneSpec) OwnerResourceName() string {
+	return ""
+}
+
+// ResourceGroupName returns the name of the resource group of the private dns zone.
+func (z ZoneSpec) ResourceGroupName() string {
+	return z.ResourceGroup
+}
+
+// Parameters returns the parameters for the private dns zone.
+func (z ZoneSpec) Parameters(existing interface{}) (params interface{}, err error) {
+	_, log, done := tele.StartSpanWithLogger(context.TODO(), "privatedns.ZoneSpec.Parameters")
+	defer done()
+
+	if existing != nil {
+		zone, ok := existing.(privatedns.PrivateZone)
+		if !ok {
+			return nil, errors.Errorf("%T is not a privatedns.PrivateZone", existing)
+		}
+		if !isManaged(zone.Tags, z.ClusterName) {
+			log.V(1).Info("Skipping reconciliation of unmanaged private DNS zone", "private DNS", zone.Name)
```

Comment:
We could add a TODO to remove this log in a future release of CAPZ. This is only a temporary warning due to #1791 adding tags previously created clusters not having the tags so this manual step is needed once for backwards compatibility.